File: app/src/main/java/com/nimtego/plectrum/presentation/main/artists/ArtistTabsFragment.java
package com.nimtego.plectrum.presentation.main.artists;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.nimtego.plectrum.App;
import com.nimtego.plectrum.presentation.main.adapter.SpacesItemDecoration;
import com.nimtego.plectrum.presentation.main.fragments.MainTabsFragment;
import com.nimtego.plectrum.presentation.main.model.ArtistModel;

import java.util.ArrayList;
import java.util.Collection;

@Deprecated
public class ArtistTabsFragment
        extends MainTabsFragment<ArtistContract.Presenter>
        implements ArtistContract.View<ArtistContract.Presenter> {

    public static ArtistTabsFragment getInstance(String response) {
        final ArtistTabsFragment fragment = new ArtistTabsFragment();
        final Bundle arguments = new Bundle();
        arguments.putString(RESPONSE, response);
        fragment.setArguments(arguments);
        return fragment;
    }


    @Override
    protected RecyclerView.ItemDecoration itemDecorator() {
        return new SpacesItemDecoration(2,
                30,
                true);
    }

    @Override
    protected RecyclerView.LayoutManager rvLayoutManager(Context context) {
        return new GridLayoutManager(context, 2);
    }

    @Override
    public void render(Collection<ArtistModel> artistModels) {
        ArtistAdapter adapter = new ArtistAdapter(new ArrayList<>(artistModels),
                this.getActivity());
        adapter.setOnItemClickListener(new ArtistAdapter.OnItemClickListener() {
            @Override
            public void onUserItemClicked(ArtistModel artistModel) {
                mPresenter.artistClicked(artistModel);
            }
        });
        mRecyclerView.setAdapter(adapter);
    }

    @Override
    public void search(String response) {
        mPresenter.search(response);
    }

    @Override
    public ArtistContract.Presenter supplyPresenter() {
        return App.getComponent().mainArtistPresenter();
    }
}

File: app/src/main/java/com/nimtego/plectrum/presentation/main/artists/ArtistTabsFragment.kt
package com.nimtego.plectrum.presentation.main.artists

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.nimtego.plectrum.presentation.main.fragments.MainTabsFragment
import com.nimtego.plectrum.presentation.main.model.ArtistModel
import com.nimtego.plectrum.presentation.ui.widget.SpaceItemDecorator

class ArtistTabsFragment : MainTabsFragment(), ArtistTabView {

    @InjectPresenter
    internal lateinit var presenter: ArtistPresenterImpl

    @ProvidePresenter
    fun provideRepositoryPresenter(): ArtistPresenterImpl {
        return ArtistPresenterImpl()
    }

    override fun render(artistModels: Collection<ArtistModel>) {
        val adapter = ArtistAdapter(ArrayList(artistModels),
                this.activity)
        adapter.setOnItemClickListener( object : ArtistAdapter.OnItemClickListener {
            override fun onUserItemClicked(artistModel: ArtistModel) {
                presenter.itemClick(artistModel)
            }
        })
        mRecyclerView?.adapter = adapter
    }

    override fun itemDecorator(): RecyclerView.ItemDecoration {
        return SpaceItemDecorator(2,
                30,
                true)
    }

    override fun rvLayoutManager(context: Context): RecyclerView.LayoutManager {
        return GridLayoutManager(context, 2)
    }

    override fun search(response: String) {
        presenter.search(response)
    }

    companion object {

        fun getInstance(response: String): ArtistTabsFragment {
            val fragment = ArtistTabsFragment()
            val arguments = Bundle()
            arguments.putString(RESPONSE, response)
            fragment.arguments = arguments
            return fragment
        }
    }
}
